content/blog/open-sourcing-workshops-on-demand-part3-understanding-the-backend.md

```diff
@@ -0,0 +1,340 @@
+
+L﻿et's see what is happening on the backend server to perform this **CLEANUP** scenario.
+
+![](/img/wod-blogserie3-cleanup.png "backend server CLEANUP workflow")
+
+A﻿s you can see, it does not differ much from the **CREATE**. We still need to gather data to interact with the proper workshop from the right student. The .procmail.rc is providing us these infos. Then, the automation kicks in through `procmail-action-sh` script.
+
+T﻿he verb is now **CLEANUP**. As a consequence, step4 is now **CLEANUP**.
+
+N﻿othing changes from 5 to 9.
+
+1﻿0- `get_wod_completion_ratio()` allows us to retrieve through a simple computing of the numbers of notebooks cells executed thoughout the different exercices of the workshop a ratio. This enables us to see how much of the workshop is actually run. Participants are asked to fill out a form in a conclusion notebook which is present in every student's workshop's folder. But only 10 % of participant do fill it. That leaves many participants for which we don't know whether they liked or not or more importantly, did they actually run it ? 
+
+T﻿his completion ratio script provides us this data and we store it in our database. 
+
+1﻿1- API call to send the completion ratio figure to the database. This can be later queried to build up nice reporting dashboard as explained here by my colleague Didier Lalli in the following [article](https://developer.hpe.com/blog/open-source-elasticsearch-helped-us-globally-support-virtual-labs/).
+
+1﻿2- `erase-student()`: now that we have extracted the necessary data from the student's notebooks, we can perform a cleanup of the student folder.
+
+1﻿3- `cleanup_processes_student()`: On top of cleanup of the student folder, we also kill all the allocated student's processes.
+
+1﻿4- `cleanup-<workshop>.sh`: If any appliance is involved, this task will perform the necessary cleanup processes on the appliance.
+
+1﻿5- Finally, just like for the creation of a workshop process, we need to inform back the frontend that now cleanup is done. Therefore, several API calls are made to update tables in the database. The new s﻿tudent password is recorded. We also generate a new password at the cleanup phase to prevent unregistered logins. The s﻿tudent status is set to inactive. The capacity figure is incremented by one to make the seat available again. 
+
+A﻿s for the **CREATE** phase, the regular checks occuring on the frontent web portal will get these data and trigger the final email to the participant thanking him for his particpation.
+
+N﻿ow let's look at the **RESET** scenario.
+
+![](/img/wod-blogserie3-reset.png "backend server RESET workflow")
+
+ You may wonder what are the differences between **CLEANUP** and **RESET**? Well, firstly, they spell differently but that has nothing to do with the purpose of this article...Secondly, **CLEANUP** only takes care of student whereas **RESET** takes care of a larger scope. Let me explain.
```

Comment:
Cute joke, but not really appropriate. If you had jokes throughout the post and the series, I'd say "go with it". But just inserting one now takes away from the flow. I'd recommend rephrasing this paragraph to:
"You may wonder about the differences between CLEANUP and RESET. CLEANUP only takes care of students whereas RESET takes care of a larger scope. Let me explain..." (yes, you can use the elipsis here)

Reply:
done